src/web/source.rs

```diff
@@ -19,42 +21,94 @@ enum FileType {
     RustSource,
 }
 
+/// A source file
 #[derive(PartialEq, PartialOrd)]
 struct File {
     name: String,
     file_type: FileType,
 }
 
+/// A list of source files
+///
+///
+/// Rust:
+///
+/// ```ignore
+/// FileList {
+///     metadata: MetaData {
+///         name: "rcc",
+///         version: "0.0.0",
+///         description: Some("it compiles an unholy language"),
+///         target_name: None,
+///         rustdoc_status: true,
+///         default_target: "x86_64-unknown-linux-gnu",
+///     },
+///     files: vec![
+///         File {
+///             name: "main.rs",
+///             file_type: FileType::RustSource,
+///         },
+///         File {
+///             name: "lib.rs",
+///             file_type: FileType::RustSource,
+///         },
+///     ],
+/// }
+/// ```
+///
+/// Json:
+///
+/// ```json
+/// {
+///     "metadata": {
+///         "name": "rcc",
+///         "version": "0.0.0",
+///         "description": "it compiles an unholy language",
+///         "target_name": null,
+///         "rustdoc_status": true,
+///         "default_target": "x86_64-unknown-linux-gnu",
+///     },
+///     "files": [{
+///         "name": "main.rs",
+///         "file_type_rust_source": true,
+///     }, {
+///         "name": "lib.rs",
+///         "file_type_rust_source": true,
+///     }],
+/// }
+/// ```
+///
 struct FileList {
     metadata: MetaData,
     files: Vec<File>,
 }
 
-impl ToJson for FileList {
-    fn to_json(&self) -> Json {
-        let mut m: BTreeMap<String, Json> = BTreeMap::new();
-
-        m.insert("metadata".to_string(), self.metadata.to_json());
-
-        let mut file_vec: Vec<Json> = Vec::new();
+impl Serialize for FileList {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("FileList", 2)?;
+        state.serialize_field("metadata", &self.metadata)?;
 
+        let mut files = Vec::with_capacity(self.files.len());
         for file in &self.files {
-            let mut file_m: BTreeMap<String, Json> = BTreeMap::new();
-            file_m.insert("name".to_string(), file.name.to_json());
+            let mut map = HashMap::with_capacity(3);
```

Comment:
We're only serializing `name` and `file_type`.
```suggestion
            let mut map = HashMap::with_capacity(2);
```

Reply:
Oops, my bad